File: apps/web/src/lib/despesas.ts
import type { AppTransporte, CategoriaDespesa, ModalidadeDespesa } from "@/lib/types";
import { formatPlaca } from "@/lib/veiculo-placa";

export const MODALIDADE_DESPESA_OPCOES: {
  value: ModalidadeDespesa;
  label: string;
  descricao: string;
}[] = [
  {
    value: "mobilidade",
    label: "Mobilidade",
    descricao: "Combustível e transporte por aplicativo",
  },
  {
    value: "livre",
    label: "Livre",
    descricao:
      "Ferramentas, materiais elétricos, consumíveis e material de escritório",
  },
];

export const CATEGORIA_DESPESA_OPCOES: {
  value: CategoriaDespesa;
  modalidade: ModalidadeDespesa;
  label: string;
}[] = [
  { value: "combustivel", modalidade: "mobilidade", label: "Combustível" },
  {
    value: "transporte_app",
    modalidade: "mobilidade",
    label: "Transporte por aplicativo",
  },
  { value: "ferramentas", modalidade: "livre", label: "Ferramentas" },
  {
    value: "materiais",
    modalidade: "livre",
    label: "Materiais (elétricos, etc.)",
  },
  {
    value: "consumiveis",
    modalidade: "livre",
    label: "Consumíveis (Café, água, etc.)",
  },
];

export function labelModalidadeDespesa(m: ModalidadeDespesa): string {
  return MODALIDADE_DESPESA_OPCOES.find((o) => o.value === m)?.label ?? m;
}

export function labelCategoriaDespesa(c: CategoriaDespesa): string {
  return CATEGORIA_DESPESA_OPCOES.find((o) => o.value === c)?.label ?? c;
}

export function categoriasPorModalidade(
  modalidade: ModalidadeDespesa
): typeof CATEGORIA_DESPESA_OPCOES {
  return CATEGORIA_DESPESA_OPCOES.filter((o) => o.modalidade === modalidade);
}

export function formatarMoeda(valor: number): string {
  return valor.toLocaleString("pt-BR", {
    style: "currency",
    currency: "BRL",
  });
}

export const APP_TRANSPORTE_OPCOES: {
  value: AppTransporte;
  label: string;
}[] = [
  { value: "uber", label: "Uber" },
  { value: "99", label: "99" },
  { value: "indrive", label: "InDrive" },
  { value: "outro", label: "Outro" },
];

export function labelAppTransporte(app?: AppTransporte, outro?: string): string {
  if (!app) return "—";
  if (app === "outro") return outro?.trim() || "Outro";
  return APP_TRANSPORTE_OPCOES.find((o) => o.value === app)?.label ?? app;
}

export function formatarDataDespesaBr(value: string): string {
  const [a, m, d] = value.split("-");
  if (a && m && d) return `${d}/${m}/${a}`;
  return value;
}

export function detalhesDespesaTexto(d: {
  descricao?: string;
  categoria: CategoriaDespesa;
  hora?: string;
  placa?: string;
  hodometro?: number;
  veiculoPessoal?: boolean;
  appTransporte?: AppTransporte;
  appTransporteOutro?: string;
}): string {
  const partes: string[] = [];
  if (d.descricao?.trim()) partes.push(d.descricao.trim());
  if (d.categoria === "combustivel") {
    if (d.hora) partes.push(d.hora);
    if (d.placa) partes.push(formatPlaca(d.placa));
    if (d.hodometro) partes.push(`${d.hodometro} km`);
    if (d.veiculoPessoal) partes.push("Veículo pessoal");
  }
  if (d.categoria === "transporte_app") {
    partes.push(labelAppTransporte(d.appTransporte, d.appTransporteOutro));
  }
  return partes.length ? partes.join(" · ") : "—";
}

File: services/api/internal/domain/despesas.go
package domain

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ModalidadeDespesa string

const (
	DespesaMobilidade ModalidadeDespesa = "mobilidade"
	DespesaLivre      ModalidadeDespesa = "livre"
)

type CategoriaDespesa string

const (
	CategoriaCombustivel     CategoriaDespesa = "combustivel"
	CategoriaTransporteApp   CategoriaDespesa = "transporte_app"
	CategoriaFerramentas     CategoriaDespesa = "ferramentas"
	CategoriaMateriais       CategoriaDespesa = "materiais"
	CategoriaConsumiveis     CategoriaDespesa = "consumiveis"
)

type AppTransporte string

const (
	AppTransporteUber    AppTransporte = "uber"
	AppTransporte99      AppTransporte = "99"
	AppTransporteIndrive AppTransporte = "indrive"
	AppTransporteOutro   AppTransporte = "outro"
)

func AppTransporteValido(a AppTransporte) bool {
	switch a {
	case AppTransporteUber, AppTransporte99, AppTransporteIndrive, AppTransporteOutro:
		return true
	default:
		return false
	}
}

func CategoriaValidaParaModalidade(m ModalidadeDespesa, c CategoriaDespesa) bool {
	switch m {
	case DespesaMobilidade:
		return c == CategoriaCombustivel || c == CategoriaTransporteApp
	case DespesaLivre:
		return c == CategoriaFerramentas || c == CategoriaMateriais || c == CategoriaConsumiveis
	default:
		return false
	}
}

// Despesa é um gasto registrado pelo colaborador (verba mobilidade ou livre).
type Despesa struct {
	ID             primitive.ObjectID  `json:"id" bson:"_id,omitempty"`
	ColaboradorID  primitive.ObjectID  `json:"colaboradorId" bson:"colaboradorId"`
	Modalidade     ModalidadeDespesa   `json:"modalidade" bson:"modalidade"`
	Categoria      CategoriaDespesa    `json:"categoria" bson:"categoria"`
	Valor          float64             `json:"valor" bson:"valor"`
	Data           string              `json:"data" bson:"data"` // yyyy-mm-dd
	Competencia    string              `json:"competencia" bson:"competencia"` // yyyy-mm
	Descricao      string              `json:"descricao,omitempty" bson:"descricao,omitempty"`
	ComprovanteURL string              `json:"comprovanteUrl,omitempty" bson:"comprovanteUrl,omitempty"`
	// Campos adicionais para categoria combustível.
	Hora           string              `json:"hora,omitempty" bson:"hora,omitempty"` // HH:mm
	VeiculoID      *primitive.ObjectID `json:"veiculoId,omitempty" bson:"veiculoId,omitempty"`
	VeiculoPessoal bool                `json:"veiculoPessoal,omitempty" bson:"veiculoPessoal,omitempty"`
	Placa          string              `json:"placa,omitempty" bson:"placa,omitempty"`
	Hodometro          int                 `json:"hodometro,omitempty" bson:"hodometro,omitempty"`
	// Campos adicionais para transporte por aplicativo.
	AppTransporte      AppTransporte       `json:"appTransporte,omitempty" bson:"appTransporte,omitempty"`
	AppTransporteOutro string              `json:"appTransporteOutro,omitempty" bson:"appTransporteOutro,omitempty"`
	CreatedAt          time.Time           `json:"createdAt" bson:"createdAt"`
	UpdatedAt      time.Time           `json:"updatedAt" bson:"updatedAt"`
}

// DepositoDespesa é o crédito mensal da empresa por modalidade.
type DepositoDespesa struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	ColaboradorID primitive.ObjectID `json:"colaboradorId" bson:"colaboradorId"`
	Competencia   string             `json:"competencia" bson:"competencia"` // yyyy-mm
	Modalidade    ModalidadeDespesa  `json:"modalidade" bson:"modalidade"`
	Valor         float64            `json:"valor" bson:"valor"`
	CreatedAt     time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt     time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// AjusteSaldoDespesa redefine o saldo a partir de um momento; movimentações anteriores
// deixam de compor o acumulado e só entram recargas/despesas posteriores ao ajuste.
type AjusteSaldoDespesa struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	ColaboradorID primitive.ObjectID `json:"colaboradorId" bson:"colaboradorId"`
	Competencia   string             `json:"competencia" bson:"competencia"` // yyyy-mm
	Modalidade    ModalidadeDespesa  `json:"modalidade" bson:"modalidade"`
	Saldo         float64            `json:"saldo" bson:"saldo"`
	AjustadoEm    time.Time          `json:"ajustadoEm" bson:"ajustadoEm"`
	AjustadoPor   string             `json:"ajustadoPor,omitempty" bson:"ajustadoPor,omitempty"`
	Observacao    string             `json:"observacao,omitempty" bson:"observacao,omitempty"`
	CreatedAt     time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt     time.Time          `json:"updatedAt" bson:"updatedAt"`
}
